File: lib/neatjson.rb
require 'json'
module JSON
	# Generate the JSON string representation for an object,
	# with a variety of formatting options.
	#
	# @author Gavin Kistner <!@phrogz.net>
	# @param object [Object] the object to serialize
	# @param opts [Hash] the formatting options
	# @option opts [Integer] :wrap            (80)    The maximum line width before wrapping. Use `false` to never wrap, or `true` to always wrap.
	# @option opts [String]  :indent          ("  ")  Whitespace used to indent each level when wrapping (without the :short option).
	# @option opts [Boolean] :indent_last     (false) Indent the closing bracket for arrays and objects (without the :short option).
	# @option opts [Boolean] :short           (false) Keep the output 'short' when wrapping, putting opening brackets on the same line as the first value, and closing brackets on the same line as the last item.
	# @option opts [Boolean] :sort            (false) Sort the keys for objects to be in alphabetical order (`true`), or supply a lambda to determine ordering.
	# @option opts [Boolean] :aligned         (false) When wrapping objects, align the colons (only per object).
	# @option opts [Boolean] :force_floats    (false) Ensure that all integer values have `.0` after them.
	# @option opts [Array]   :force_floats_in ([])    Array of object key names to force floats inside of.
	# @option opts [Integer] :decimals        (null) Decimal precision to use for floats; omit to keep numeric values precise.
	# @option opts [Boolean] :trim_trailing_zeros (false) Remove trailing zeros added by `decimals`` from the ends of floating point numbers.
	# @option opts [Integer] :padding         (0) Number of spaces to put inside brackets/braces for both arrays and objects.
	# @option opts [Integer] :array_padding   (0) Number of spaces to put inside brackets for arrays. Overrides `:padding`.
	# @option opts [Integer] :object_padding  (0) Number of spaces to put inside braces for objects. Overrides `:padding`.
	# @option opts [Integer] :around_comma    (0) Number of spaces to put before/after commas (for both arrays and objects).
	# @option opts [Integer] :before_comma    (0) Number of spaces to put before commas (for both arrays and objects).
	# @option opts [Integer] :after_comma     (0) Number of spaces to put after commas (for both arrays and objects).
	# @option opts [Integer] :around_colon    (0) Number of spaces to put before/after colons (for objects).
	# @option opts [Integer] :before_colon    (0) Number of spaces to put before colons (for objects).
	# @option opts [Integer] :after_colon     (0) Number of spaces to put after colons (for objects).
	# @option opts [Integer] :around_colon_1  (0) Number of spaces to put before/after colons for single-line objects.
	# @option opts [Integer] :before_colon_1  (0) Number of spaces to put before colons for single-line objects.
	# @option opts [Integer] :after_colon_1   (0) Number of spaces to put after colons for single-line objects.
	# @option opts [Integer] :around_colon_n  (0) Number of spaces to put before/after colons for multi-line objects.
	# @option opts [Integer] :before_colon_n  (0) Number of spaces to put before colons for multi-line objects.
	# @option opts [Integer] :after_colon_n   (0) Number of spaces to put after colons for multi-line objects.
	# @return [String] the JSON representation of the object.
	#
	# The lambda for the `sort` option will be passed the string name of the key, the value, and the hash for the object being sorted.
	# The values returned for all keys must be all comparable, or an error will occur.
	def self.neat_generate(object,opts={})
		opts ||= {}
		opts[:wrap] = 80 unless opts.key?(:wrap)
		opts[:wrap] = -1 if opts[:wrap]==true
		opts[:indent]         ||= "  "
		opts[:force_floats_in]||= []
		opts[:array_padding]  ||= opts[:padding]      || 0
		opts[:object_padding] ||= opts[:padding]      || 0
		opts[:after_comma]    ||= opts[:around_comma] || 0
		opts[:before_comma]   ||= opts[:around_comma] || 0
		opts[:before_colon]   ||= opts[:around_colon] || 0
		opts[:after_colon]    ||= opts[:around_colon] || 0
		opts[:before_colon_1] ||= opts[:around_colon_1] || opts[:before_colon] || 0
		opts[:after_colon_1]  ||= opts[:around_colon_1] || opts[:after_colon]  || 0
		opts[:before_colon_n] ||= opts[:around_colon_n] || opts[:before_colon] || 0
		opts[:after_colon_n]  ||= opts[:around_colon_n] || opts[:after_colon]  || 0
		raise ":indent option must only be whitespace" if opts[:indent]=~/\S/

		apad  = " " * opts[:array_padding]
		opad  = " " * opts[:object_padding]
		comma = "#{' '*opts[:before_comma]},#{' '*opts[:after_comma]}"
		colon1= "#{' '*opts[:before_colon_1]}:#{' '*opts[:after_colon_1]}"
		colonn= "#{' '*opts[:before_colon_n]}:#{' '*opts[:after_colon_n]}"

		memoizer = {}
		build = ->(o,indent,floats_forced) do
			memoizer[[o,indent,floats_forced]] ||= case o
				when String               then "#{indent}#{self.generate(o)}"
				when Symbol               then "#{indent}#{self.generate(o.to_s)}"
				when TrueClass,FalseClass then "#{indent}#{o}"
				when NilClass             then "#{indent}null"
				when Integer
					floats_forced ? build[o.to_f, indent, floats_forced] : "#{indent}#{o.inspect}"
				when Float
					if o.infinite?
						"#{indent}#{o<0 ? "-9e9999" : "9e9999"}"
					elsif o.nan?
						"#{indent}\"NaN\""
					elsif !floats_forced && (o==o.to_i) && (o.to_s !~ /e/)
						build[o.to_i, indent, floats_forced]
					elsif opts[:decimals]
						if opts[:trim_trailing_zeros]
							"#{indent}#{o.round(opts[:decimals])}"
						else
							"#{indent}%.#{opts[:decimals]}f" % o
						end
					else
						"#{indent}#{o}"
					end

				when Array
					if o.empty?
						"#{indent}[]"
					else
						pieces = o.map{ |v| build[v, '', floats_forced] }
						one_line = "#{indent}[#{apad}#{pieces.join comma}#{apad}]"
						if !opts[:wrap] || (one_line.length <= opts[:wrap])
							one_line
						elsif opts[:short]
							indent2 = "#{indent} #{apad}"
							pieces = o.map{ |v| build[v, indent2, floats_forced] }
							pieces[0] = pieces[0].sub indent2, "#{indent}[#{apad}"
							pieces[pieces.length-1] = "#{pieces.last}#{apad}]"
							pieces.join ",\n"
						else
							indent2 = "#{indent}#{opts[:indent]}"
							"#{indent}[\n#{o.map{ |v| build[v, indent2, floats_forced] }.join ",\n"}\n#{opts[:indent_last] ? indent2 : indent}]"
						end
					end

				when Hash
					if o.empty?
						"#{indent}{}"
					else
						case sort=(opts[:sorted] || opts[:sort])
							when true then o = o.sort_by{|k,v| k.to_s }
							when Proc
								o = case sort.arity
								when 1 then o.sort_by{ |k,v| sort[k] }
								when 2 then o.sort_by{ |k,v| sort[k,v] }
								when 3 then o.sort_by{ |k,v| sort[k,v,o] }
								end
						end
						keys = o.map{ |x| x.first.to_s }
						keyvals = o.map.with_index{ |(k,v),i| [ self.generate(k.to_s), build[v, '', opts[:force_floats] || opts[:force_floats_in].include?(keys[i])] ] }
						keyvals = keyvals.map{ |kv| kv.join(colon1) }.join(comma)
						one_line = "#{indent}{#{opad}#{keyvals}#{opad}}"
						if !opts[:wrap] || (one_line.length <= opts[:wrap])
							one_line
						else
							if opts[:short]
								keyvals = o.map{ |k,v| ["#{indent} #{opad}#{self.generate(k.to_s)}",v] }
								keyvals[0][0] = keyvals[0][0].sub "#{indent} ", "#{indent}{"
								if opts[:aligned]
									longest = keyvals.map(&:first).map(&:length).max
									formatk = "%-#{longest}s"
									keyvals.map!{ |k,v| [ formatk % k,v] }
								end
								keyvals.map!.with_index do |(k,v),i|
									floats_forced = opts[:force_floats] || opts[:force_floats_in].include?(keys[i])
									indent2 = " "*"#{k}#{colonn}".length
									one_line = "#{k}#{colonn}#{build[v, '', floats_forced]}"
									if opts[:wrap] && (one_line.length > opts[:wrap]) && (v.is_a?(Array) || v.is_a?(Hash))
										"#{k}#{colonn}#{build[v, indent2, floats_forced].lstrip}"
									else
										one_line
									end
								end
								"#{keyvals.join(",\n")}#{opad}}"
							else
								keyvals = o.map{ |k,v| ["#{indent}#{opts[:indent]}#{self.generate(k.to_s)}",v] }
								if opts[:aligned]
									longest = keyvals.map(&:first).map(&:length).max
									formatk = "%-#{longest}s"
									keyvals.map!{ |k,v| [ formatk % k,v] }
								end
								indent2 = "#{indent}#{opts[:indent]}"
								keyvals.map!.with_index do |(k,v),i|
									floats_forced = opts[:force_floats] || opts[:force_floats_in].include?(keys[i])
									one_line = "#{k}#{colonn}#{build[v, '', floats_forced]}"
									if opts[:wrap] && (one_line.length > opts[:wrap]) && (v.is_a?(Array) || v.is_a?(Hash))
										"#{k}#{colonn}#{build[v, indent2, floats_forced].lstrip}"
									else
										one_line
									end
								end
								"#{indent}{\n#{keyvals.join(",\n")}\n#{opts[:indent_last] ? indent2 : indent}}"
							end
						end
					end

				else
					"#{indent}#{o.to_json(opts)}"
			end
		end

		build[object, '', opts[:force_floats]]
	end
end

File: javascript/neatjson.js
(function(exports){
exports.neatJSON = neatJSON;

function neatJSON(value,opts){
	opts = opts || {}
	if (!('wrap'          in opts)) opts.wrap = 80;
	if (opts.wrap==true) opts.wrap = -1;
	if (!('indent'        in opts)) opts.indent = '  ';
	if (!('arrayPadding'  in opts)) opts.arrayPadding  = ('padding' in opts) ? opts.padding : 0;
	if (!('objectPadding' in opts)) opts.objectPadding = ('padding' in opts) ? opts.padding : 0;
	if (!('beforeComma'   in opts)) opts.beforeComma   = ('aroundComma' in opts) ? opts.aroundComma : 0;
	if (!('afterComma'    in opts)) opts.afterComma    = ('aroundComma' in opts) ? opts.aroundComma : 0;
	if (!('beforeColon'   in opts)) opts.beforeColon   = ('aroundColon' in opts) ? opts.aroundColon : 0;
	if (!('afterColon'    in opts)) opts.afterColon    = ('aroundColon' in opts) ? opts.aroundColon : 0;
	if (!('beforeColon1'  in opts)) opts.beforeColon1  = ('aroundColon1' in opts) ? opts.aroundColon1 : ('beforeColon' in opts) ? opts.beforeColon : 0;
	if (!('afterColon1'   in opts)) opts.afterColon1   = ('aroundColon1' in opts) ? opts.aroundColon1 : ('afterColon'  in opts) ? opts.afterColon  : 0;
	if (!('beforeColonN'  in opts)) opts.beforeColonN  = ('aroundColonN' in opts) ? opts.aroundColonN : ('beforeColon' in opts) ? opts.beforeColon : 0;
	if (!('afterColonN'   in opts)) opts.afterColonN   = ('aroundColonN' in opts) ? opts.aroundColonN : ('afterColon'  in opts) ? opts.afterColon  : 0;
	if (!('forceFloatsIn' in opts)) opts.forceFloatsIn = [];

	const apad   = repeat(' ',opts.arrayPadding),
	      opad   = repeat(' ',opts.objectPadding),
	      comma  = repeat(' ',opts.beforeComma)+','+repeat(' ',opts.afterComma),
	      colon1 = repeat(' ',opts.beforeColon1)+':'+repeat(' ',opts.afterColon1),
	      colonN = repeat(' ',opts.beforeColonN)+':'+repeat(' ',opts.afterColonN);

	const build = memoize();
	return build(value,'',opts.forceFloats);

	function memoize(){
		const memo = new Map;
		return function(o,indent,floatsForced){
			const byFloats = floatsForced ? {o,floatsForced} : o;
			let byIndent=memo.get(byFloats);
			if (!byIndent) memo.set(byFloats,byIndent={});
			if (!byIndent[indent]) byIndent[indent] = rawBuild(o,indent,floatsForced);
			return byIndent[indent];
		}
	}

	function rawBuild(o,indent,floatsForced){
		if (o===null || o===undefined) return indent+'null';
		else{
			if (typeof o==='number'){
				if (o===Infinity) {
					return indent+'9e9999';
				}else if (o===-Infinity){
					return indent+'-9e9999';
				}else if (Number.isNaN(o)){
					return indent+'"NaN"';
				}else{
					const treatAsFloat = floatsForced || (o === +o && o !== (o|0));
					let result = indent + ((treatAsFloat && ('decimals' in opts)) ? o.toFixed(opts.decimals) : (o+''));
					if (opts.trimTrailingZeros) result = (+result)+'';
					if (floatsForced && result.indexOf('.')==-1) result += '.0';
					return result;
				}
			}else if (o instanceof Array){
				if (!o.length) return indent+"[]";
				let pieces = o.map(function(v){ return build(v,'',floatsForced) });
				const oneLine = indent+'['+apad+pieces.join(comma)+apad+']';
				if (opts.wrap===false || oneLine.length<=opts.wrap) return oneLine;
				if (opts.short){
					const indent2 = indent+' '+apad;
					pieces = o.map(function(v){ return build(v,indent2,floatsForced) });
					pieces[0] = pieces[0].replace(indent2,indent+'['+apad);
					pieces[pieces.length-1] = pieces[pieces.length-1]+apad+']';
					return pieces.join(',\n');
				}else{
					const indent2 = indent+opts.indent;
					return indent+'[\n'+o.map(function(v){ return build(v,indent2,floatsForced) }).join(',\n')+'\n'+(opts.indentLast?indent2:indent)+']';
				}
			}else if (o instanceof Object){
				let sortedKV=[], i=0;
				const sort = opts.sort || opts.sorted;
				for (let k in o){
					const kv = sortedKV[i++] = [k,o[k]];
					if (sort===true) kv[2] = k;
					else if (typeof sort==='function') kv[2]=sort(k,o[k],o);
				}
				if (!sortedKV.length) return indent+'{}';
				if (sort) sortedKV = sortedKV.sort(function(a,b){ a=a[2]; b=b[2]; return a<b?-1:a>b?1:0 });
				let keyvals=sortedKV.map(([k,v]) => [JSON.stringify(k), build(v,'',(opts.forceFloats || opts.forceFloatsIn.includes(k)))].join(colon1) ).join(comma);
				const oneLine = indent+"{"+opad+keyvals+opad+"}";
				if (opts.wrap===false || oneLine.length<opts.wrap) return oneLine;
				if (opts.short){
					keyvals = sortedKV.map(function(kv){ return [indent+' '+opad+JSON.stringify(kv[0]), kv[1]] });
					keyvals[0][0] = keyvals[0][0].replace(indent+' ',indent+'{');
					if (opts.aligned){
						let longest = 0;
						for (let i=keyvals.length;i--;) if (keyvals[i][0].length>longest) longest = keyvals[i][0].length;
						const padding = repeat(' ',longest);
						for (let i=keyvals.length;i--;) keyvals[i][0] = padRight(padding,keyvals[i][0]);
					}
					for (let i=keyvals.length;i--;){
						let k=keyvals[i][0], v=keyvals[i][1];
						const indent2 = repeat(' ',(k+colonN).length);
						floatsForced = (opts.forceFloats || opts.forceFloatsIn.includes(k));
						const oneLine = k+colonN+build(v,'',floatsForced);
						keyvals[i] = (opts.wrap===false || oneLine.length<=opts.wrap || !v || typeof v!="object") ? oneLine : (k+colonN+build(v,indent2,floatsForced).replace(/^\s+/,''));
					}
					return keyvals.join(',\n') + opad + '}';
				}else{
					const keyvals=sortedKV.map(function(kvs){ kvs[0] = indent+opts.indent+JSON.stringify(kvs[0]); return kvs });
					if (opts.aligned){
						let longest = 0;
						for (let i=keyvals.length;i--;) if (keyvals[i][0].length>longest) longest = keyvals[i][0].length;
						const padding = repeat(' ',longest);
						for (let i=keyvals.length;i--;) keyvals[i][0] = padRight(padding,keyvals[i][0]);
					}
					const indent2 = indent+opts.indent;
					for (let i=keyvals.length;i--;){
						const k=keyvals[i][0], v=keyvals[i][1];
						const oneLine = k+colonN+build(v,'',floatsForced);
						keyvals[i] = (opts.wrap===false || oneLine.length<=opts.wrap || !v || typeof v!="object") ? oneLine : (k+colonN+build(v,indent2,floatsForced).replace(/^\s+/,''));
					}
					return indent+'{\n'+keyvals.join(',\n')+'\n'+(opts.indentLast?indent2:indent)+'}'
				}
			}else{
				return indent+JSON.stringify(o);
			}
		}
	}

	function repeat(str,times){ // http://stackoverflow.com/a/17800645/405017
		let result = '';
		while(true){
			if (times & 1) result += str;
			times >>= 1;
			if (times) str += str;
			else break;
		}
		return result;
	}

	function padRight(pad, str){
		return (str + pad).substring(0, pad.length);
	}
}
neatJSON.version = "0.10";

})(typeof exports === 'undefined' ? this : exports);
